src/state/image-dataset/csv-dataset/index.ts

```diff
@@ -0,0 +1,398 @@
+            if (!indexInfo) {
+                // Assign new index to this value
+                indexInfo = { index: strValueToIndex.size, count: 0 };
+                strValueToIndex.set(value, indexInfo);
+            }
+
+            indexInfo.count++;
+            remappedValues.push(indexInfo.index);
+        }
+
+        const options: Record<string, MeasuredFeaturesOption> = {};
+        for (const [value, { index, count }] of strValueToIndex.entries()) {
+            options[index.toString()] = {
+                color: DEFAULT_COLORS[index % DEFAULT_COLORS.length],
+                name: value,
+                key: value,
+                count: count,
+            };
+        }
+
+        return {
+            def: {
+                discrete: true,
+                displayName: key,
+                description: key,
+                key,
+                options,
+                tooltip: key,
+            },
+            data: remappedValues,
+        };
+    }
+
+    private parseFeatures(csvData: Record<string, string>[]): void {
+        this.featureInfo.clear();
+
+        const featureKeys = this.getFeatureKeysFromColumnNames(csvData);
+        const rawFeatureData = this.getFeatureDataAsColumns(csvData, featureKeys);
+
+        for (const key of featureKeys) {
+            const data = rawFeatureData.get(key);
+            if (!data) {
+                continue;
+            }
+            if (isStringArray(data)) {
+                const { def, data: discreteData } = this.parseDiscreteFeature(key, data);
+                this.featureInfo.set(key, {
+                    type: FeatureType.DISCRETE,
+                    def,
+                    data: discreteData,
+                });
+            } else {
+                const def: ContinuousMeasuredFeatureDef = {
+                    discrete: false,
+                    displayName: key,
+                    description: key,
+                    key,
+                    tooltip: key,
+                };
+                this.featureInfo.set(key, {
+                    type: FeatureType.CONTINUOUS,
+                    def,
+                    data: data,
+                });
+            }
+        }
+
+        // TODO: Feature defs can include units. Should we strip that from the feature column name?
+    }
+
+    private parseCsvData(csvDataSrc: string): void {
+        // TODO: handle URLs and files here: they need to be handled via async callbacks.
+        // https://www.papaparse.com/docs#strings
+        const result = Papa.parse(csvDataSrc, { header: true }).data as Record<string, string>[];
+        this.csvData = result as Record<string, string>[];
+
+        if (this.csvData.length === 0) {
+            throw new Error("No data found in CSV");
+        }
+
+        // Map from cell IDs to row index. If no cell ID is provided, assign the row number.
+        for (let i = 0; i < this.csvData.length; i++) {
```

Comment:
So does this allow a mix of cellid values and nulls?  If so, doesn't that allow for duplicate values?  EG,
row, cellid
0, myCellId
1, 2
2,nan

Reply:
Good point, it would... I'd need to do a sweep through the whole column to make sure `CellId` exists for all of them, hmm. Let me tinker with this one!